### PR TITLE
Adding explicit cast to expression in nested initializer list

### LINKED ISSUE #295: Dredd mutates expressions in nested initializer list
Applying dredd to
```
struct foo {
    short x;
    short y;
};

int main() {
    foo bar[] = {{1, 2}, {3, 4}};
}
```
mutate the intializer list to
```
{{__dredd_replace_expr_int_one(1, 0), __dredd_replace_expr_int_constant(2, 3)}, {__dredd_replace_expr_int_constant(3, 8), __dredd_replace_expr_int_constant(4, 13)}};
```
which result in compile-time error
```
error: non-constant-expression cannot be narrowed from type 'int' to 'short' in initializer list [-Wc++11-narrowing]
```

The nested initialiser list bypass fix #278

### PR DESCRIPTION
**Pull request by JonathanFoo0523**
Up until now, an explicit `static_cast` has been applied to expressions
under one layer of initializer list in case Dredd introduces a mutator
function that narrows the type of the expression. This PR extends the
feature so that an explicit `static_cast` is applied to expressions
under nested initializer lists.

Fixes #295

### CODE REVIEW

File: src/libdredd/src/mutation_replace_expr.cc [hunk 1]
```diff
@@ -584,10 +584,10 @@ void MutationReplaceExpr::ReplaceExprWithFunctionCall(
   // cast.
   auto parents = ast_context.getParents<clang::Expr>(*expr_);
   // This the expression occurs in an initializer list and is subject to an
-  // explicit cast then it will have two parents -- the initializer list, and
-  // the implicit cast. (This is probably due to implicit casts being treated
-  // as invisible nodes in the AST.)
-  if (ast_context.getLangOpts().CPlusPlus && parents.size() == 2 &&
+  // explicit cast then it will have at least two parents -- the initializer
```
Comment (afd) [inline src/libdredd/src/mutation_replace_expr.cc:587]: In the above line please change "This the expression" to "This expression" (existing wording issue of mine that I noticed while reviewing this).

File: test/single_file/initializer_list_narrower_nested.cc [hunk 2]
```diff
@@ -0,0 +1,8 @@
+struct foo {
+  short x;
+  short y;
+};
+
+void baz() {
+  foo bar[] = {{1, 2}};
+}
```
Comment (afd) [inline test/single_file/initializer_list_narrower_nested.cc:8]: Please add new line at end of file.

File: test/single_file/initializer_list_narrower_nested.cc.expected [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+struct foo {
+  short x;
+  short y;
+};
+
+void baz() {
+  foo bar[] = {{static_cast<short>(__dredd_replace_expr_int_one(1, 0)), static_cast<short>(__dredd_replace_expr_int_constant(2, 3))}};
+}
```
Comment (afd) [inline test/single_file/initializer_list_narrower_nested.cc.expected:70]: Ditto.

File: test/single_file/initializer_list_narrower_nested.cc.noopt.expected [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+struct foo {
+  short x;
+  short y;
+};
+
+void baz() {
+  foo bar[] = {{static_cast<short>(__dredd_replace_expr_int(1, 0)), static_cast<short>(__dredd_replace_expr_int(2, 6))}};
+}
```
Comment (afd) [inline test/single_file/initializer_list_narrower_nested.cc.noopt.expected:63]: Ditto.

**Review by afd — APPROVED** [review]
Looks good - please merge after making these minor edits.

File: src/libdredd/src/mutation_replace_expr.cc (hunk 1 above)
Comment (JonathanFoo0523) [reply src/libdredd/src/mutation_replace_expr.cc:587]: Rephrased to
"The expression occurs in an initializer list and is then subject to an explicit cast will have at least two parents -- the initializer list, and the implicit cast. "

File: test/single_file/initializer_list_narrower_nested.cc (hunk 2 above)
Comment (JonathanFoo0523) [reply test/single_file/initializer_list_narrower_nested.cc:8]: fixed

File: test/single_file/initializer_list_narrower_nested.cc.expected (hunk 3 above)
Comment (JonathanFoo0523) [reply test/single_file/initializer_list_narrower_nested.cc.expected:70]: regenerated

File: test/single_file/initializer_list_narrower_nested.cc.noopt.expected (hunk 4 above)
Comment (JonathanFoo0523) [reply test/single_file/initializer_list_narrower_nested.cc.noopt.expected:63]: regenerated

**Pull request merged**